src/LodeRunner.API.Test/UnitTests/IntervalCheckerTest.cs

```diff
@@ -0,0 +1,155 @@
+            using var loggerFactory = LoggerFactory.Create(logger =>
+            {
+                logger.Setup(logLevelConfig: config, logValues: config, projectName: projectName);
+            });
+
+            return loggerFactory.CreateLogger<IntervalCheckerTest>();
+        }
+
+        /// <summary>
+        /// Calculate the maximum waiting time.
+        /// </summary>
+        /// <param name="interval">The interval.</param>
+        /// <returns>the max waiting time.</returns>
+        private static int CalculateTimeout(int interval)
+        {
+            int newInterval = interval;
+
+            for (int i = 1; i < 3; i++)
+            {
+                newInterval /= 2;
+                interval += newInterval;
+            }
+
+            // We will wait for Max Interval time plus 1 extra second.
+            return interval + 1;
+        }
+
+        /// <summary>
+        /// Runs the interval checker and waits until timeout and get cancellation token status and console output.
+        /// </summary>
+        /// <param name="taskToExecute">The task to execute.</param>
+        /// <param name="intervalSeconds">The interval seconds.</param>
+        /// <param name="retryLimit">The retry limit.</param>
+        /// <returns>The cancellationToken status and the outputAsStringList.</returns>
+        private async Task<bool> RunIntervalCheckerWaitAndGetCancellationStatus(Func<Task<bool>> taskToExecute, int intervalSeconds = 4, int retryLimit = 3)
+        {
+            CancellationTokenSource cancellationTokenSource = new ();
+
+            using var intervalChecker = new IntervalChecker(taskToExecute, this.logger, cancellationTokenSource, intervalSeconds, retryLimit);
+
+            intervalChecker.Start();
+
+            return await this.WaitForTimeoutAndGetCancellationRequestStatus(cancellationTokenSource, intervalSeconds);
+        }
+
+        /// <summary>
+        /// Waits until timeout is reached and determines if Cancellation request is in place.
+        /// </summary>
+        /// <param name="cancellationTokenSource">The cancellation token source.</param>
+        /// <param name="intervalSeconds">The interval seconds.</param>
+        /// <returns>Whether the cancellation was requested or not.</returns>
+        private async Task<bool> WaitForTimeoutAndGetCancellationRequestStatus(CancellationTokenSource cancellationTokenSource, int intervalSeconds)
+        {
+            int maxWaitingTime = CalculateTimeout(intervalSeconds);
+            await Task.Delay(maxWaitingTime * 1000).ConfigureAwait(false);
+
+            bool cancellationRequested = cancellationTokenSource.IsCancellationRequested;
+            this.output.WriteLine($"Total Wait time: {maxWaitingTime} secs.\tCancellation Requested: {cancellationRequested}\t");
+            return cancellationRequested;
+        }
+
+        /// <summary>
+        /// Gets a boolean true.
```

Comment:
Add some more description as to why this is defined this way.  When I first looked at this I was wondering why you wrapped bool response in a Task<>.   I then realized that you need the task as the method being called takes an async function pointer.  So, please describe why this exists a little better in the comments.

Reply:
Added more description to the new refactored method.